node/src/services/forest_storage.rs

```diff
@@ -256,4 +302,42 @@ where
     async fn remove_forest_storage(&mut self, key: &Self::Key) {
         self.fs_instances.write().await.remove(key);
     }
+
+    async fn snapshot(
+        &self,
+        src_key: &Self::Key,
+        dest_key: &Self::Key,
+    ) -> Option<Arc<RwLock<Self::FS>>> {
+        let mut fs_instances = self.fs_instances.write().await;
+
+        // Return potentially existing instance since we waited for the lock.
+        // This is for the case where many threads called `snapshot` at the same time with the same `dest_key`.
+        if let Some(fs) = fs_instances.get(dest_key) {
+            return Some(fs.clone());
+        }
+
+        let storage_path = self
+            .storage_path
+            .clone()
+            .expect("Storage path should be set");
+        let src = format!("{}_{:?}", storage_path, src_key);
+        let dest = format!("{}_{:?}", storage_path, dest_key);
+
+        let underlying_db = match rocksdb::copy_db(src, dest) {
```

Comment:
I'm pretty sure this is not 100% correct and that it will be very hard for it to surface. I suggest we guard lock operations on the `src` in order to guarantee no writes from other threads.